// beliefs/dist/claude-agent-sdk.js
// src/claude-agent-sdk.ts
function beliefsHooks(beliefs, options = {}) {
  const {
    capture = "tools",
    includeContext = true,
    toolFilter
  } = options;
  const hooks = {};
  if (includeContext) {
    hooks.SessionStart = [
      {
        hooks: [
          async (_input) => {
            try {
              const context = await beliefs.before();
              return {
                continue: true,
                hookSpecificOutput: {
                  hookEventName: "SessionStart",
                  additionalContext: context.prompt
                }
              };
            } catch (err) {
              console.error("[beliefs/claude-agent-sdk] SessionStart hook failed:", err);
              return { continue: true };
            }
          }
        ],
        timeout: 30
      }
    ];
  }
  if (capture === "tools" || capture === "all") {
    hooks.PostToolUse = [
      {
        matcher: toolFilter,
        hooks: [
          async (input) => {
            try {
              const postInput = input;
              const toolName = postInput.tool_name;
              const toolResponse = postInput.tool_response;
              if (toolResponse != null) {
                const text = typeof toolResponse === "string" ? toolResponse : JSON.stringify(toolResponse);
                await beliefs.after(text, { tool: toolName });
              }
            } catch (err) {
              console.error("[beliefs/claude-agent-sdk] PostToolUse hook failed:", err);
            }
            return { continue: true };
          }
        ],
        timeout: 30
      }
    ];
  }
  return hooks;
}
export {
  beliefsHooks
};
