// beliefs/dist/claude-agent-sdk.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/claude-agent-sdk.ts
var claude_agent_sdk_exports = {};
__export(claude_agent_sdk_exports, {
  beliefsHooks: () => beliefsHooks
});
module.exports = __toCommonJS(claude_agent_sdk_exports);
function beliefsHooks(beliefs, options = {}) {
  const {
    capture = "tools",
    includeContext = true,
    toolFilter
  } = options;
  const hooks = {};
  if (includeContext) {
    hooks.SessionStart = [
      {
        hooks: [
          async (_input) => {
            try {
              const context = await beliefs.before();
              return {
                continue: true,
                hookSpecificOutput: {
                  hookEventName: "SessionStart",
                  additionalContext: context.prompt
                }
              };
            } catch (err) {
              console.error("[beliefs/claude-agent-sdk] SessionStart hook failed:", err);
              return { continue: true };
            }
          }
        ],
        timeout: 30
      }
    ];
  }
  if (capture === "tools" || capture === "all") {
    hooks.PostToolUse = [
      {
        matcher: toolFilter,
        hooks: [
          async (input) => {
            try {
              const postInput = input;
              const toolName = postInput.tool_name;
              const toolResponse = postInput.tool_response;
              if (toolResponse != null) {
                const text = typeof toolResponse === "string" ? toolResponse : JSON.stringify(toolResponse);
                await beliefs.after(text, { tool: toolName });
              }
            } catch (err) {
              console.error("[beliefs/claude-agent-sdk] PostToolUse hook failed:", err);
            }
            return { continue: true };
          }
        ],
        timeout: 30
      }
    ];
  }
  return hooks;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  beliefsHooks
});
