// beliefs/dist/index.js
// src/errors.ts
var BETA_MESSAGE = "beliefs is in private beta. Request access at https://thinkn.ai/waitlist or email devs@thinkn.ai";
var BetaAccessError = class extends Error {
  code = "BETA_ACCESS_REQUIRED";
  signupUrl = "https://thinkn.ai/waitlist";
  constructor(message) {
    super(message ?? BETA_MESSAGE);
    this.name = "BetaAccessError";
  }
};
var BeliefsError = class extends Error {
  code;
  retryable;
  retryAfterMs;
  constructor(code, message, retryable = false, retryAfterMs) {
    super(message);
    this.name = "BeliefsError";
    this.code = code;
    this.retryable = retryable;
    this.retryAfterMs = retryAfterMs;
  }
};
function parseServerError(status, body) {
  const code = body.code ?? "internal/error";
  const message = body.message ?? `HTTP ${status}`;
  if (status === 401 || status === 403) {
    return new BetaAccessError(message);
  }
  if (status === 429) {
    return new BeliefsError(code, message, true);
  }
  if (status === 400) {
    return new BeliefsError(code, message, false);
  }
  return new BeliefsError(code, message, status >= 500);
}

// src/schemas/scope.ts
import { z } from "zod";
var BeliefsConfigSchema = z.object({
  apiKey: z.string(),
  agent: z.string().optional(),
  namespace: z.string().optional(),
  thread: z.string().optional(),
  baseUrl: z.string().optional(),
  timeout: z.number().positive().optional(),
  maxRetries: z.number().int().min(0).optional(),
  debug: z.boolean().optional()
});
var ScopeSchema = z.object({
  namespace: z.string().optional(),
  thread: z.string().optional(),
  threadId: z.string().optional()
});

// src/version.ts
var VERSION = "0.1.0";

// src/client.ts
var DEFAULT_BASE_URL = "https://www.thinkn.ai";
var DEFAULT_TIMEOUT_MS = 12e4;
var Beliefs = class {
  apiKey;
  agent;
  namespace;
  thread;
  baseUrl;
  timeout;
  maxRetries;
  debug;
  constructor(config) {
    const parsed = BeliefsConfigSchema.parse(config);
    this.apiKey = parsed.apiKey;
    this.agent = parsed.agent ?? "agent";
    this.namespace = parsed.namespace ?? "default";
    this.thread = parsed.thread;
    this.baseUrl = parsed.baseUrl ?? DEFAULT_BASE_URL;
    this.timeout = parsed.timeout ?? DEFAULT_TIMEOUT_MS;
    this.maxRetries = parsed.maxRetries ?? 2;
    this.debug = parsed.debug ?? false;
  }
  /**
   * Before the agent acts — get context, clarity, and suggested next moves.
   *
   * Inject `context.prompt` into your agent's system prompt.
   */
  async before(input) {
    const res = await this._post("/api/sdk/v1/beliefs/context", {
      ...this._scope(),
      input
    });
    return {
      prompt: res.system ?? JSON.stringify(res.context ?? {}),
      beliefs: (res.snapshot?.beliefs ?? []).map(mapBelief),
      goals: res.context?.goals ?? [],
      gaps: (res.snapshot?.gaps ?? []).map((g) => g.text ?? ""),
      clarity: res.clarity?.score ?? 0,
      channels: mapChannels(res.clarity?.channels),
      moves: (res.moves ?? []).map(mapMove)
    };
  }
  /**
   * After the agent acts — feed the observation.
   *
   * Returns what changed and the full updated world state.
   */
  async after(text, options) {
    if (options?.tool) {
      const res2 = await this._post("/api/sdk/v1/beliefs/ingest-tool-result", {
        ...this._scope(),
        toolName: options.tool,
        toolResult: text,
        ...options.source ? { source: options.source } : {}
      });
      return mapDelta(res2);
    }
    const res = await this._post("/api/sdk/v1/beliefs/ingest", {
      ...this._scope(),
      output: text,
      ...options?.source ? { source: options.source } : {}
    });
    return mapDelta(res);
  }
  async add(textOrItems, options) {
    if (Array.isArray(textOrItems)) {
      return this._addMany(textOrItems);
    }
    const type = options?.type ?? "claim";
    const body = { ...this._scope() };
    if (type === "goal") {
      body.goals = [textOrItems];
    } else if (type === "gap") {
      body.gaps = [textOrItems];
    } else {
      const claim = {
        text: textOrItems,
        confidence: options?.confidence ?? 0.5,
        type
      };
      if (options?.source) claim.source = options.source;
      body.claims = [claim];
      if (options?.evidence) {
        body.evidence = [{ content: options.evidence, supports: textOrItems }];
      }
      if (options?.supersedes) {
        body.gapsResolved = [options.supersedes];
      }
    }
    const res = await this._post("/api/sdk/v1/beliefs/ingest", body);
    return mapDelta(res);
  }
  /**
   * @deprecated Use `add(items)` instead. This method will be removed in a future version.
   */
  async addMany(items) {
    return this._addMany(items);
  }
  async _addMany(items) {
    const body = { ...this._scope() };
    const claims = [];
    const goals = [];
    const gaps = [];
    const evidence = [];
    for (const item of items) {
      const type = item.type ?? "claim";
      if (type === "goal") {
        goals.push(item.text);
      } else if (type === "gap") {
        gaps.push(item.text);
      } else {
        const claim = {
          text: item.text,
          confidence: item.confidence ?? 0.5,
          type
        };
        if (item.source) claim.source = item.source;
        claims.push(claim);
        if (item.evidence) {
          evidence.push({ content: item.evidence, supports: item.text });
        }
      }
    }
    if (claims.length > 0) body.claims = claims;
    if (goals.length > 0) body.goals = goals;
    if (gaps.length > 0) body.gaps = gaps;
    if (evidence.length > 0) body.evidence = evidence;
    const res = await this._post("/api/sdk/v1/beliefs/ingest", body);
    return mapDelta(res);
  }
  /**
   * Mark a gap as resolved.
   */
  async resolve(text) {
    const res = await this._post("/api/sdk/v1/beliefs/ingest", {
      ...this._scope(),
      gapsResolved: [text]
    });
    return mapDelta(res);
  }
  /**
   * Read the full fused world state.
   */
  async read() {
    const res = await this._post("/api/sdk/v1/beliefs/context", {
      ...this._scope()
    });
    const beliefs = (res.snapshot?.beliefs ?? []).map(mapBelief);
    const edges = (res.snapshot?.edges ?? []).map(mapEdge);
    return {
      beliefs,
      goals: res.context?.goals ?? [],
      gaps: (res.snapshot?.gaps ?? []).map((g) => g.text ?? ""),
      edges,
      contradictions: res.snapshot?.contradictions ?? [],
      clarity: res.clarity?.score ?? 0,
      channels: mapChannels(res.clarity?.channels),
      moves: (res.moves ?? []).map(mapMove),
      prompt: String(res.system ?? JSON.stringify(res.context ?? {}))
    };
  }
  /**
   * Read a lightweight snapshot of beliefs without computing clarity or moves.
   * Faster than `read()` for cases where you only need the current state.
   */
  async snapshot() {
    const params = new URLSearchParams({ namespace: this.namespace });
    if (this.thread) params.set("thread", this.thread);
    const res = await this._get(`/api/sdk/v1/beliefs/snapshot?${params}`);
    const data = res;
    return {
      beliefs: (data.beliefs ?? []).map(mapBelief),
      goals: (data.goals ?? []).map((g) => g.text ?? ""),
      gaps: (data.gaps ?? []).map((g) => g.text ?? ""),
      edges: (data.edges ?? []).map(mapEdge),
      contradictions: data.contradictions ?? []
    };
  }
  /**
   * Retract a belief — marks it as retracted.
   * The belief stays in the graph with lifecycle: 'retracted'.
   * Use this when the agent no longer believes something.
   */
  async retract(beliefId, reason) {
    const res = await this._post("/api/sdk/v1/beliefs/apply", {
      ...this._scope(),
      delta: {
        retract: [{ beliefId, ...reason ? { reason } : {} }]
      }
    });
    return mapDelta(res);
  }
  /**
   * Remove a belief — deletes it from the graph entirely.
   * A final ledger entry is recorded for traceability.
   * Use this for cleanup of garbage or accidental beliefs.
   */
  async remove(beliefId) {
    const res = await this._post("/api/sdk/v1/beliefs/apply", {
      ...this._scope(),
      delta: {
        remove: [beliefId]
      }
    });
    return mapDelta(res);
  }
  /**
   * Reset — remove all beliefs, goals, gaps, and intents in this scope.
   * Every removal is recorded in the ledger for traceability.
   */
  async reset() {
    return this._post("/api/sdk/v1/beliefs/reset", this._scope());
  }
  /**
   * Search beliefs by text.
   */
  async search(query) {
    const params = new URLSearchParams({ namespace: this.namespace, query });
    if (this.thread) params.set("thread", this.thread);
    const res = await this._request("GET", `/api/sdk/v1/beliefs/search?${params}`);
    return (Array.isArray(res) ? res : []).map(mapBelief);
  }
  /**
   * Trace belief evolution history.
   */
  async trace(beliefId) {
    const params = new URLSearchParams({ namespace: this.namespace });
    if (this.thread) params.set("thread", this.thread);
    if (beliefId) params.set("beliefId", beliefId);
    const res = await this._get(`/api/sdk/v1/beliefs/ledger?${params}`);
    return (res.entries ?? []).map((e) => mapTraceEntry(e));
  }
  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------
  _scope() {
    const scope = {
      namespace: this.namespace,
      agentId: this.agent
    };
    if (this.thread) scope.thread = this.thread;
    return scope;
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async _post(path, body) {
    return this._request("POST", path, body);
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async _get(path) {
    return this._request("GET", path);
  }
  async _request(method, path, body) {
    if (!this.apiKey) throw new BetaAccessError();
    const url = `${this.baseUrl}${path}`;
    if (this.debug) {
      console.log(`[beliefs] ${method} ${path}`, body ? JSON.stringify(body).slice(0, 200) : "");
    }
    let lastError;
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.timeout);
      try {
        const res = await fetch(url, {
          method,
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            "User-Agent": `beliefs-sdk/${VERSION}`,
            ...body ? { "Content-Type": "application/json" } : {}
          },
          body: body ? JSON.stringify(body) : void 0,
          signal: controller.signal
        });
        if (!res.ok) {
          const data = await res.json().catch(() => ({}));
          throw parseServerError(res.status, data);
        }
        const json = await res.json();
        if (this.debug) {
          console.log(`[beliefs] ${method} ${path} \u2192 ${res.status}`);
        }
        return json;
      } catch (err) {
        lastError = err;
        const isRetryableServer = err instanceof BeliefsError && err.retryable;
        const isNetworkError = err instanceof TypeError || err instanceof DOMException && err.name === "AbortError";
        const isRetryable = isRetryableServer || isNetworkError;
        if (!isRetryable || attempt >= this.maxRetries) throw err;
        const baseDelay = Math.min(1e3 * 2 ** attempt, 1e4);
        const delay = (err instanceof BeliefsError ? err.retryAfterMs : void 0) ?? baseDelay;
        if (this.debug) {
          console.log(`[beliefs] retry ${attempt + 1}/${this.maxRetries} in ${delay}ms`);
        }
        await sleep(delay);
        continue;
      } finally {
        clearTimeout(timer);
      }
    }
    throw lastError;
  }
};
function mapDelta(res) {
  const clarityScore = res.clarity?.score ?? 0;
  const readinessRaw = res.clarity?.readiness;
  let readiness = "low";
  if (readinessRaw === "high" || readinessRaw === "medium" || readinessRaw === "low") {
    readiness = readinessRaw;
  } else if (clarityScore >= 0.7) {
    readiness = "high";
  } else if (clarityScore >= 0.4) {
    readiness = "medium";
  }
  const snapshot = res.snapshot;
  const beliefs = (snapshot?.beliefs ?? []).map(mapBelief);
  const edges = (snapshot?.edges ?? []).map(mapEdge);
  const gaps = (snapshot?.gaps ?? []).map((g) => g.text ?? "");
  const contradictions = snapshot?.contradictions ?? [];
  const goals = res.context ? res.context?.goals ?? [] : [];
  const channels = mapChannels(res.clarity?.channels);
  const state = {
    beliefs,
    goals,
    gaps,
    edges,
    contradictions,
    clarity: clarityScore,
    channels,
    moves: (res.moves ?? []).map(mapMove),
    prompt: String(res.system ?? JSON.stringify(res.context ?? {}))
  };
  const changes = (res.ledgerEntries ?? []).map(mapChange);
  return {
    changes,
    clarity: clarityScore,
    channels,
    readiness,
    moves: state.moves,
    state
  };
}
function mapChange(e) {
  const before = e.before;
  const after = e.after;
  return {
    action: e.action ?? "created",
    beliefId: String(e.beliefId ?? ""),
    text: String(e.text ?? ""),
    confidence: before || after ? {
      before: before?.confidence,
      after: after?.confidence
    } : void 0,
    reason: e.reasoning ? String(e.reasoning) : void 0,
    source: e.source ? String(e.source) : void 0
  };
}
function mapBelief(b) {
  return {
    id: String(b.id ?? ""),
    text: String(b.text ?? ""),
    confidence: Number(b.confidence ?? 0.5),
    type: String(b.type ?? b.semanticLabel ?? "claim"),
    label: b.semanticLabel ?? void 0,
    evidenceWeight: typeof b.evidenceWeight === "number" ? b.evidenceWeight : void 0,
    distribution: isDistributionKind(b.distribution) ? b.distribution : void 0,
    lifecycle: typeof b.lifecycle === "string" ? b.lifecycle : void 0,
    provenance: typeof b.provenance === "string" ? b.provenance : void 0,
    source: typeof b.source === "string" ? b.source : void 0,
    createdAt: String(b.createdAt ?? (/* @__PURE__ */ new Date()).toISOString()),
    updatedAt: b.updatedAt ? String(b.updatedAt) : void 0
  };
}
function isDistributionKind(v) {
  return v === "claim" || v === "category" || v === "measurement";
}
function mapMove(m) {
  return {
    action: String(m.type ?? m.suggestedAction ?? "research"),
    target: String(m.target ?? ""),
    targetId: m.targetId ? String(m.targetId) : void 0,
    reason: String(m.reasoning ?? ""),
    value: Number(m.expectedInfoGain ?? 0),
    executor: m.executor ?? void 0
  };
}
function mapEdge(e) {
  return {
    type: String(e.type ?? "supports"),
    source: String(e.source ?? ""),
    target: String(e.target ?? ""),
    confidence: Number(e.confidence ?? 0)
  };
}
function mapTraceEntry(e) {
  return {
    action: e.action ?? "created",
    beliefId: e.beliefId ? String(e.beliefId) : void 0,
    confidence: e.before || e.after ? {
      before: e.before?.confidence,
      after: e.after?.confidence
    } : void 0,
    agent: e.agentId ? String(e.agentId) : void 0,
    source: e.source ? String(e.source) : void 0,
    timestamp: String(e.timestamp ?? (/* @__PURE__ */ new Date()).toISOString()),
    reason: e.reasoning ? String(e.reasoning) : void 0
  };
}
function mapChannels(raw) {
  if (!raw || typeof raw !== "object") return void 0;
  const c = raw;
  if (typeof c.decisionResolution !== "number" && typeof c.knowledgeCertainty !== "number" && typeof c.coherence !== "number" && typeof c.coverage !== "number") return void 0;
  return {
    decisionResolution: Number(c.decisionResolution ?? 0),
    knowledgeCertainty: Number(c.knowledgeCertainty ?? 0),
    coherence: Number(c.coherence ?? 0),
    coverage: Number(c.coverage ?? 0)
  };
}
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// src/beta-access.ts
var WAITLIST_URL = "https://www.thinkn.ai/api/waitlist";
async function requestBetaAccess(request) {
  const res = await fetch(WAITLIST_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      email: request.email,
      name: request.name,
      source: "sdk",
      context: request.context
    })
  });
  if (!res.ok) {
    const data2 = await res.json().catch(() => ({}));
    return {
      success: false,
      message: data2.error || "Request failed. Please email devs@thinkn.ai directly."
    };
  }
  const data = await res.json();
  return data;
}
export {
  Beliefs,
  BeliefsConfigSchema,
  BeliefsError,
  BetaAccessError,
  Beliefs as beliefs,
  Beliefs as default,
  requestBetaAccess
};
