// beliefs/dist/vercel-ai.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/vercel-ai.ts
var vercel_ai_exports = {};
__export(vercel_ai_exports, {
  beliefsMiddleware: () => beliefsMiddleware
});
module.exports = __toCommonJS(vercel_ai_exports);
function beliefsMiddleware(beliefs, options = {}) {
  const {
    capture = "response",
    includeContext = true
  } = options;
  return {
    specificationVersion: "v3",
    transformParams: async ({
      params,
      type: _type
    }) => {
      if (!includeContext) return params;
      const userMessage = extractLastUserMessage(params);
      const context = await beliefs.before(userMessage);
      return injectContext(params, context);
    },
    wrapGenerate: async ({
      doGenerate,
      params
    }) => {
      const result = await doGenerate();
      if (capture === "response" || capture === "all") {
        const text = extractGeneratedText(result);
        if (text) {
          await beliefs.after(text).catch(logError);
        }
      }
      if (capture === "tools" || capture === "all") {
        await extractAndProcessToolResults(result, beliefs);
      }
      return result;
    },
    wrapStream: async ({
      doStream
    }) => {
      const result = await doStream();
      const originalStream = result.stream;
      if (!originalStream) return result;
      const textChunks = [];
      const toolResults = [];
      const transformedStream = originalStream.pipeThrough(
        new TransformStream({
          transform(chunk, controller) {
            controller.enqueue(chunk);
            const chunkObj = chunk;
            if (chunkObj.type === "text-delta" && typeof chunkObj.delta === "string") {
              textChunks.push(chunkObj.delta);
            }
            if (chunkObj.type === "tool-result") {
              toolResults.push({
                toolName: String(chunkObj.toolName ?? "unknown"),
                result: chunkObj.result
              });
            }
          },
          async flush() {
            try {
              const pending = [];
              if ((capture === "tools" || capture === "all") && toolResults.length > 0) {
                for (const tr of toolResults) {
                  pending.push(beliefs.after(JSON.stringify(tr.result), { tool: tr.toolName }).catch(logError));
                }
              }
              const fullText = textChunks.join("");
              if ((capture === "response" || capture === "all") && fullText) {
                pending.push(beliefs.after(fullText).catch(logError));
              }
              await Promise.all(pending);
            } catch (err) {
              logError(err);
            }
          }
        })
      );
      return { ...result, stream: transformedStream };
    }
  };
}
function extractLastUserMessage(params) {
  const prompt = params.prompt;
  if (!Array.isArray(prompt)) return void 0;
  for (let i = prompt.length - 1; i >= 0; i--) {
    const msg = prompt[i];
    if (msg.role === "user") {
      const content = msg.content;
      if (typeof content === "string") return content;
      if (Array.isArray(content)) {
        const textPart = content.find(
          (p) => p.type === "text"
        );
        return textPart?.text;
      }
    }
  }
  return void 0;
}
function injectContext(params, context) {
  const prompt = params.prompt;
  if (!Array.isArray(prompt)) return params;
  const systemIdx = prompt.findIndex((m) => m.role === "system");
  const contextText = context.prompt;
  if (systemIdx >= 0) {
    const existing = prompt[systemIdx];
    const existingContent = typeof existing.content === "string" ? existing.content : "";
    const updatedPrompt = [...prompt];
    updatedPrompt[systemIdx] = {
      ...existing,
      content: existingContent + "\n\n" + contextText
    };
    return { ...params, prompt: updatedPrompt };
  }
  return {
    ...params,
    prompt: [
      { role: "system", content: contextText },
      ...prompt
    ]
  };
}
function extractGeneratedText(result) {
  if (typeof result.text === "string") return result.text;
  const content = result.content;
  if (Array.isArray(content)) {
    const texts = content.filter((c) => c.type === "text").map((c) => String(c.text ?? ""));
    if (texts.length > 0) return texts.join("");
  }
  return void 0;
}
async function extractAndProcessToolResults(result, beliefs) {
  const content = result.content;
  if (!Array.isArray(content)) return;
  const toolBlocks = content.filter(
    (block) => block.type === "tool-result" && block.result != null
  );
  await Promise.all(
    toolBlocks.map((block) => {
      const toolName = String(block.toolName ?? "unknown");
      return beliefs.after(JSON.stringify(block.result), { tool: toolName }).catch(logError);
    })
  );
}
function logError(err) {
  console.error("[beliefs/vercel-ai] Error processing beliefs:", err);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  beliefsMiddleware
});
